### ISSUE TITLE
Safer deploy of updates

### ISSUE THREAD

**Issue by Sjors** [bitcoin-dot-org/bitcoin.org#1939]
The particular change discussed in #1938 was merely controversial, but if any of the maintainers Github accounts is ever compromised, that could be used to put malware on a site that's pretty high in the Google rankings.

It's possible to use Travis with Github Pages: https://docs.travis-ci.com/user/deployment/pages/

Travis could check GPG signatures of authors. It could also require some threshold of maintainers to sign off on changes, though I'm not sure how one would implement the latter.

Of course anyone with merge permissions could just change the Travis script and signature whitelist, but at least this would be very obvious and easy to monitor.

**Comment by wbnns** [comment]
@Sjors Agreed. We used to have this but unfortunately @Cobra-Bitcoin turned it off earlier this year, despite objections; here's the last update we received from him on it:

> Currently it's set up to just build locally from the repository if there's any new commits, without any of the signature checking (that was becoming a real hassle, and even if there's some sort of malicious attack through Github, someone would notice it quickly anyway). There's no longer any need for the build machine too, since this server just does the whole build process locally using a slightly modified version of the build script.

I'm assigning this issue to him, since he's the only one with control over the hosting environment.

**Comment by Sjors** [comment]
I didn't realize you're not using Github pages.

Another approach, which the community can do themselves, is to have script which compares the contents of bitcoin.org to what's in the repo. That script can also check if due process was followed, using whatever standard the script author deems right. This would also catch a DNS hijacking scenario.

I agree iwht @Cobra-Bitcoin that any issue would be detected quite quickly (as we've seen yesterday), but there's something to be said for situational awareness, so people don't need to guess if an contributor went rogue (or duress) or the site is under attack.

**Comment by harding** [comment]
@Sjors If you checkout the repository, build a local copy of the site with `make`, and then run `make manual-check-diff-sha256sums`, you'll get a diff of the self-reported sha256sums between the server version of the site and your local version of the site.  You can then manually review the differing files.

This was designed for discovering bugs, not for security, so it can't guarantee detection (nothing can, actually, since an attacker can serve the expected file to Inspector Alice but a malicious file to User Bob).

**Comment by kuzzmi** [comment]
Why can't we just make the whole thing automated so that we can reduce amount of human activity to minimum? E.g. the website can be deployed only through a workflow of a PR being merged to master?

What are the downsides to that?

I've done that several times in more or less similar scenarios, so that nobody could just redeploy the whole thing manually without letting the rest of the team know about that.

Let's say we use Travis not only as a test runner, but also for deployments, then we could make a user on a server that can do just one thing - deploy the website, and that users' key can be stored in Travis (they support encrypting one file and putting it in the repo). In that scenario nobody but those who have root access to the machine or Travis can redeploy the website.

If anybody finds this idea bad, please let me know why.

**Comment by harding** [comment]
@kuzzmi That's what they had, and it sounds like they probably still have that but without the authorized-signature-required security.  See https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/_build/update_site.sh for the old automatic build script.
